text/0000-symbol-name-mangling-v2.md

```diff
@@ -0,0 +1,801 @@
+
+<path-root> := <crate-id>
+            |  M <type>
+            |  X <type> <abs-path>
+```
+
+Here are some examples for complete symbol names:
+
+```
+<u32 as mycrate::Foo>::foo => _RNXmN12mycrate_abcd3FooE3fooVE
+mycrate::Foo<u32>::foo => _RNMN12mycrate_abcd3FooImEE3fooVE
+<mycrate::Foo<u32> as mycrate::Bar<u64>>::foo => _RNXN12mycrate_abcd3FooImEEN12mycrate_abcd3BarIyEE3fooVE
+```
+
+
+### Items Within Specialized Trait Impls
+
+In Rust one can define items within generic items, e.g. functions or impls, like in the following example:
+
+```rust
+fn foo<T>(a: T) -> (u32, T) {
+  static mut X: u32 = 0;
+
+  unsafe {
+    X += 1;
+    (X, a)
+  }
+}
+```
+
+The `X` here (or any other such nested definition) does not inherit the generic context. `X` is non-generic, and a function defined in its place would be too. Consequently, when giving the path to something defined within a generic item, one does not specify the generic arguments because they add no information. The fully qualified name of `X` is thus `my_crate[a1b2c3d4]::foo::X` and its symbol name:
+
+```
+_RN15mycrate_4a3b56d3fooF1XFE
+```
+
+However, there is at least one case where the type arguments *do* matter for a definition like this, and that is when trait specialization is used. Consider the following piece of code:
+
+```rust
+trait Foo<T> {
+  fn foo() -> T;
+}
+
+struct Bar<T>(T);
+
+impl<T> Foo<T> for Bar<T> {
+  default fn foo() -> T {
+    static MSG: &str = "sry, no can do";
+    panic!("{}", MSG)
+  }
+}
+
+impl<T: Default> Foo<T> for Bar<T> {
+  fn foo() -> T {
+    static MSG: &str = "it's a go!";
+    println!("{}", MSG);
+    T::default()
+  }
+}
+
+```
+
+Notice that, if one just omits the type arguments, both `MSG` statics have the path `<Bar as Foo>::foo::MSG`. However, we cannot disambiguate by adding type arguments, because we don't have any concrete types to substitute the arguments for. Therefore, we have to disambiguate the `impls`. Since trait specialization is an unstable feature of Rust and the details are in flux, this RFC does not try to provide a mangling based on the `where` clauses of the specialized `impls`. Instead it proposes to re-use the "numeric disambiguator" form already used for macro hygiene and closures. Thus, conflicting `impls` would be disambiguated via an implementation defined suffix, as in `<Bar as Foo>'1::foo::MSG` and `<Bar as Foo>'2::foo::MSG`. This encoding introduces minimal additional syntax and can be replaced with something more human-readable once the definition of trait specialization is final.
+
+
+### Unicode Identifiers
+
+Rust allows Unicode identifiers but our character set is restricted to ASCII alphanumerics, and `_`. In order to transcode the former to the latter, we use the same approach as Swift, which is: encode all non-ascii identifiers via [Punycode][punycode], a standardized and efficient encoding that keeps encoded strings in a rather human-readable format. So for example, the string
```

Comment:
```suggestion
Rust allows Unicode identifiers but our character set is restricted to ASCII alphanumerics, and `_`. In order to transcode the former to the latter, we use the same approach as Swift, which is: encode all non-ASCII identifiers via [Punycode][punycode], a standardized and efficient encoding that keeps encoded strings in a rather human-readable format. So for example, the string
```

Reply:
>  via Punycode

Punycode is defined in such a way that no two Unicode strings map to the same output, right?

What would these two map to?

```rust
fn gödel() {}
fn göödel() {}
```

Reply:
> Punycode is defined in such a way that no two Unicode strings map to the same output, right?

I'm not sure what you're asking ("no two" *distinct* "Unicode strings"?). But for your example, this is what my implementation produces (in a crate named `foo`):
* `gödel`: `gdel_Fqa` (`_RNnCsbDqzXfLQacH_3foou8gdel_Fqa`)
* `göödel`: `gdel_Fqaa` (`_RNnCsbDqzXfLQacH_3foou9gdel_Fqaa`)

Reply:
Everything after the last `-` is a set of base36-encoded compressed instructions to insert arbitrary codepoints at arbitrary positions in the string before. (and if there's no ASCII initial string, the `-` is also missing)

There's no notion of "collision" as this is a perfectly lossless encoding, and it can't even use less than a base36 (ASCII) character to encode an Unicode character.

In my example, the second `a` in `Fqaa` is the most compressed encoding in punycode (it's `0` in their base36): "insert the same codepoint as the last one, just after the last insert position".

You can test this out on online punycode decoders, just keep in mind most of them also want a `xn--` prefix (which is what DNS uses).

Reply:
The `u` flag indicates whether punycode encoding is being used.